core/src/main/java/com/microsoft/applicationinsights/internal/shutdown/SDKShutdownActivity.java

```diff
@@ -118,25 +123,51 @@ private void stopStoppables() {
                 }
             }
         }
+        private void closeClosables() {
+            for (Closeable c : closeables) {
+                try {
+                    c.close();
+                } catch (ThreadDeath td) {
+                    throw td;
+                } catch (Throwable t) {
+                    try {
+                        InternalLogger.INSTANCE.error("Failed to close closeable class '%s': %s", c.getClass().getName(), t.toString());
+                        InternalLogger.INSTANCE.trace("Stack trace: %s", ExceptionUtils.getStackTrace(t));
+                    } catch (ThreadDeath td2) {
+                        throw td2;
+                    } catch (Throwable t2) {
+                        // chomp
+                    }
+                }
+            }
+        }
     }
 
-    private static volatile SDKShutdownThread shutdownThread;
+    private static volatile SDKShutdownAction shutdownAction;
 
     public void register(ChannelFetcher fetcher) {
-        getShutdownThread().register(fetcher);
+        getShutdownAction().register(fetcher);
     }
 
     public void register(Stoppable stoppable) {
-        getShutdownThread().register(stoppable);
+        getShutdownAction().register(stoppable);
+    }
+
+    public void register(Closeable closable) {
+        getShutdownAction().register(closable);
+    }
+
+    public void stopAll() {
+        getShutdownAction().run();
     }
 
-    private SDKShutdownThread getShutdownThread() {
-        if (shutdownThread == null) {
+    private SDKShutdownAction getShutdownAction() {
+        if (shutdownAction == null) {
             synchronized (this) {
```

Comment:
@littleaj  As a one cent advise, I would prefer to start a practice of using `Reentrant Lock` instead of synchronizing on `this`. There are numerous benefits . I would like everyone to go over. 
https://stackoverflow.com/questions/11821801/why-use-a-reentrantlock-if-one-can-use-synchronizedthis
Also it is much more cleaner and less prone to errors.

Reply:
I don't see any benefits of using `RentrantLock`. IMO, Java's lock classes are much less readable and harder to debug. I'm happy to listen to evaluate some evidence why `synchronized` is inadequate for this case.

However, (quoting the best answer from the linked stackoverflow article):
> ... use it when you actually need something it provides that synchronized doesn't ... but remember that the vast majority of synchronized blocks hardly ever exhibit any contention, let alone high contention. I would advise developing with synchronization until synchronization has proven to be inadequate, rather than simply assuming "the performance will be better" if you use ReentrantLock. Remember, these are advanced tools for advanced users. (And truly advanced users tend to prefer the simplest tools they can find until they're convinced the simple tools are inadequate.) ...

Reply:
Ok, I have no objections. I personally feel writing synchronized is two step process whereas reentrant lock just makes it more readable and easy.